text/0000-generalized-type-ascription.md

```diff
@@ -0,0 +1,1563 @@
+    }
+}
+```
+
+we move up on the deprecation schedule of [RFC 1685] and propose that
+writing:
+
+```rust
+trait Foo {
+    fn bar(MyType) -> ... { ... }
+}
+```
+
+will cause the compiler to emit a warn-by-default lint in Rust 2015 and
+that it be a hard error in Rust 2018.
+
+## Linting ascription of named `struct` literals and patterns
+
+Consider a struct:
+
+```rust
+struct Foo<T> {
+    bar: T
+}
+```
+
+When it comes to type ascribing the field `bar` in a struct literal expression
+such as `Foo { bar: x : Type }` or in particular when you type ascribe
+`bar` in a pattern: `Foo { bar: x : Type }` it is not always very clear from
+this way of writing what is what.
+
+We propose therefore that the compiler should provide a warn-by-default lint
+that suggests that you should wrap the ascription in parenthesis like so:
+
+```rust
+let x = Foo { bar: (x : Type) }
+
+let Foo { bar: (x: Type) } = ...;
+```
+
+This lint only applies when `x` is an identifier and not otherwise.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Grammar
+
+Assuming the following tokens and the terminal strings they lex:
+
+```rust
+LOOP : 'loop' ;
+UNSAFE : 'unsafe' ;
+TRY : 'try' ;
+ASYNC : 'async' ;
+LET : 'let' ;
+SELF : 'self' ;
+DOTDOTDOT : '...' ;
+```
+
+The following alternatives are modified in the expression grammar:
+
+```rust
+ascribe: ':' ty_sum ;
+
+expr
+: ...
+| expr ascribe // This is specified in RFC 803 but it is included for completeness.
+| LOOP ascribe? block // This replaces the existing production for loop { .. }.
+| UNSAFE ascribe? block
+| TRY ascribe? block
+| ASYNC ascribe? block
+;
+```
+
+Here, the precedence of `:` in the alternative `expr ascribe` is the same as
+the operator `as`. However, when the parser encounters type ascription of an
+expression immediately followed by a field projection or a method call,
+then the parser shall interpret the projection and the call as being performed
+on the ascribed expression. Thus, if a user writes `expr : type . field`
+the parser associates this as `(expr : type) . field`. Similarly, if a user
+writes `expr : type . method(..)` the parser associates this as
+`(expr : type) . method(..)`. An implementation of this wrt. method calls
+exists in [rust-lang/rust#33380](https://github.com/rust-lang/rust/pull/33380).
+
+To the pattern grammar, the following alternative is added:
+
+```rust
+pat
+: ...
+| pat ascribe
+;
+```
+
+The operator `:` binds more tightly than `ref`, `ref mut`, `&`, and `&mut`
+in pattern contexts. This is required because currently, the following compiles:
+
+```rust
+#[derive(Copy, Clone)]
+struct X {}
+let a = X {};
+
+let ref b: X = a; // Note the type!
+let &c : &X = b; // And here!
+```
+
+This entails for example that a Rust compiler will interpret `ref x : T` as
+`ref (x : T)` instead of `(ref x) : T`. The same applies to `ref mut`, `&`,
+and `&mut`.
+
+The grammar of `let` bindings is changed from:
```

Comment:
It automatically applies to `if let` and `while let` because they already used the `pat` grammar, see https://github.com/rust-lang/rust/blob/master/src/grammar/parser-lalr.y#L1761-L1764 (could be outdated but I don't think it is in this case).